File: java/src/main/java/org/sammancoaching/ReportStep.java
package org.sammancoaching;

import org.sammancoaching.dependencies.Config;
import org.sammancoaching.dependencies.Emailer;
import org.sammancoaching.dependencies.Logger;
import org.sammancoaching.dependencies.Project;

public record ReportStep(String name, Config config, Logger log, Emailer emailer) implements PipelineStep {
    @Override
    public PipelineStepResult run(Project project, PipelineStepResult stepResult) {
        if (config.sendEmailSummary()) {
            log.info("Sending email");
            if (stepResult.stepPassed()) {
                emailer.send("Deployment completed successfully");
            } else {
                emailer.send(stepResult.failureReason());
            }
        } else {
            log.info("Email disabled");
        }
        return new PipelineStepResult(this.name(), true, stepResult.failureReason());
    }
}

File: csharp/BuildPipeline/ReportStep.cs
namespace UntangledConditionals
{
    public class ReportStep : PipelineStep
    {
        private readonly string _name;
        private readonly Config _config;
        private readonly Logger _log;
        private readonly Emailer _emailer;

        public ReportStep(string name, Config config, Logger log, Emailer emailer)
        {
            _name = name;
            _config = config;
            _log = log;
            _emailer = emailer;
        }

        public PipelineStepResult Run(Project project, PipelineStepResult stepResult)
        {
            if (_config.SendEmailSummary())
            {
                _log.Info("Sending email");
                if (stepResult.StepPassed)
                {
                    _emailer.Send("Deployment completed successfully");
                }
                else
                {
                    _emailer.Send(stepResult.FailureReason);
                }
            }
            else
            {
                _log.Info("Email disabled");
            }

            return new PipelineStepResult(_name, true, stepResult.FailureReason);
        }
    }
}

